File: packages/pinstripe/lib/database/column.js

import { Base } from "../base.js";
import { createAdapterDeligator } from './adapter.js';

const deligateToAdapter = createAdapterDeligator('column');

export const Column = Base.extend().include({
    initialize(table, name, type){
        this._adapter = table._database._adapter;
        this._table = table;
        this._name = name;
        this._type = type;
    },

    renderSql: deligateToAdapter('renderSql'),

    toSql: deligateToAdapter('toSql'),

    async type(){
        if(!this._type){
            this._type = (await this._table[this._name])._type;
        }
        return this._type;
    },

    async exists(){
        return (await this.type()) !== undefined;
    },

    create: deligateToAdapter('create'),

    drop: deligateToAdapter('drop'),

    async __beforeInspect(){
        const exists = await this.exists();
        this._inspectInfo = {
            exists,
            ...(() => {
                if(exists){
                    return {
                        type: this._type
                    };
                }
                return {};
            })()
        };
    },

    __inspect(){
        return `${this._name} (Column) ${JSON.stringify(this._inspectInfo, null, 2)}`;
    }
});

File: projects/slick/lib/slick/database/column.rb

require "slick/database/adaptable"

class Slick::Database::Column

    class << self
        include Slick::Database::Adaptable
    end

    _define_adapter_methods :column

    deligate_to_adapter :create, :drop    

    attr_reader :name

    def initialize(table, name, type = nil)
        @table = table
        @name = name
        @type = type
    end

    def type
        @type ||= @table.send(@name).instance_variable_get(:@type)
    end

    def exist?
        !type.nil?
    end

    def _adapter
        @_adapter ||= @table.database.send(:_adapter)
    end

end
